### PR TITLE
Fixing downloading of containers if name has SHA instead of tag

### PR DESCRIPTION
**Pull request by zed-rishabh**
Containerd is unable to import the image if the image name has SHA instead of tag in the config while downloading.

### CODE REVIEW

File: pkg/pillar/zedUpload/ociutil/oci.go [hunk 1]
```diff
@@ -121,6 +122,18 @@ func Pull(registry, repo, localFile, username, apiKey string, client *http.Clien
...
 	defer w.Close()
 
+	tag, ok := ref.(name.Tag)
+	if !ok {
+		d, ok := ref.(name.Digest)
+		if !ok {
+			err := fmt.Errorf("Image name %s doesn't have a tag or digest", ref)
+			return manifestDirect, manifestResolved, size, err
+		}
+		parts := strings.Split(d.String(), ":")
```
Comment (deitch) [inline pkg/pillar/zedUpload/ociutil/oci.go:132]: No need to play with the string. You can replace all of the `parts := strings.Split(...)` with:

```go
digestTag := fmt.Sprintf("dummyTag-%s", d.DigestStr())
```

Comment (deitch) [reply pkg/pillar/zedUpload/ociutil/oci.go:132]: See https://godoc.org/github.com/google/go-containerregistry/pkg/name#Digest.DigestStr

Comment (zed-rishabh) [reply pkg/pillar/zedUpload/ociutil/oci.go:132]: Thanks. I am updating the PR.

Comment (zed-rishabh) [reply pkg/pillar/zedUpload/ociutil/oci.go]: Resolved.

**Comment by deitch** [comment]
Got some yetus and `go vet` things to fix up

**Comment by zed-rishabh** [comment]
> Got some yetus and `go vet` things to fix up

Done.

**Comment by zed-rishabh** [comment]
This may be because the corrupted image is present in verified directory which is not allowing the download of new file after fix. So, two solution to fix it. One is reboot the device and wait for 10 min for images to be garbage collected and second is delete file from the verified directory and reboot the system.

File: pkg/pillar/zedUpload/ociutil/oci.go (hunk 1 above)
Comment (zed-rishabh) [reply pkg/pillar/zedUpload/ociutil/oci.go]: @deitch After these changes, I am seeing new error ```normalize image ref "index.docker.io/library/alpine:dummyTag-sha256:a143f3ba578f79e2c7b3022c488e6e12a35836cd4a6eb9e363d7f3a07d848590": invalid reference format```. It seems like the underlying library is not happy with the format. So, I am reverting these changes.

**Comment by deitch** [comment]
That tag doesn’t look right. It looks like you still are getting “sha256:<hash>” instead of just “<hash>”. I’ll take another look shortly.

File: pkg/pillar/zedUpload/ociutil/oci.go (hunk 1 above)
Comment (deitch) [reply pkg/pillar/zedUpload/ociutil/oci.go:132]: You are correct. I must have been looking at the wrong thing. `DigestStr()` gives `sha256:<hash>`, and not just `<hash>`. You still need to split it on `:`. Sorry.

Comment (zed-rishabh) [reply pkg/pillar/zedUpload/ociutil/oci.go:132]: No issues. I updated the PR. Thanks!

**Comment by eriknordmark** [comment]
> This may be because the corrupted image is present in verified directory which is not allowing the download of new file after fix. So, two solution to fix it. One is reboot the device and wait for 10 min for images to be garbage collected and second is delete file from the verified directory and reboot the system.

When @deitch and I looked yesterday the tar file in the verified directory looked fine.
Looking right now the file is
/persist/downloads/appImg.obj/verified/1F62BE6B6EBE4EF3AB816F47C851510B1C0ABF7A42A83F01059886393A0C81D4/ignition-oxy\@sha256\:1f62be6b6ebe4ef3ab816f47c851510b1c0abf7a42a83f01059886393a0c81d4
and it contains a
imagemanifest-1f62be6b6ebe4ef3ab816f47c851510b1c0abf7a42a83f01059886393a0c81d4.json

Thus that doesn't look corrupted at all.

FWIW I added a log to show the tag, and it is
index.docker.io/mmadany/ignition-oxy:dummyTag-sha256:1f62be6b6ebe4ef3ab816f47c851510b1c0abf7a42a83f01059886393a0c81d4

Is that what you intended?

**Comment by deitch** [comment]
> it contains a
imagemanifest-1f62be6b6ebe4ef3ab816f47c851510b1c0abf7a42a83f01059886393a0c81d4.json

Is that _all_ it contains?

>  I added a log to show the tag, and it is
index.docker.io/mmadany/ignition-oxy:dummyTag-sha256:1f62be6b6ebe4ef3ab816f47c851510b1c0abf7a42a83f01059886393a0c81d4

That isn't right, and it is my fault. It should be `dummyTag-1f62be6b6ebe4ef3ab816f47c851510b1c0abf7a42a83f01059886393a0c81d4`. I think that @zed-rishabh fixed it now though.

File: pkg/pillar/zedUpload/ociutil/oci.go (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/zedUpload/ociutil/oci.go:132]: And there is always a ':' in the d.String()?

Note there is now an unused
const (
	digestTag = "digest-without-tag"
)
to be cleaned up later.

Comment (zed-rishabh) [reply pkg/pillar/zedUpload/ociutil/oci.go:132]: > And there is always a ':' in the d.String()?

Yes, d.String() will always return ```<cont-name>@sha256:<hash>```. But it makes sense to check for the length after splitting.

Comment (deitch) [reply pkg/pillar/zedUpload/ociutil/oci.go:132]: It does, but if you use `d.DigestStr()`, it _always_ will return `sha256:<hash>` (which was my mistake in the first place). But still a good idea to check length.

**Pull request merged**